src/main/java/se/liu/ida/hefquin/engine/datastructures/impl/SolutionMappingsHashTable.java

```diff
@@ -1,85 +1,209 @@
+		// TODO: is there a more efficient way to implement this method?
+		final List<SolutionMapping> solMap = new ArrayList<>();
+		final Iterator<List<SolutionMapping>> li = map.values().iterator();
+		while(li.hasNext()){
+			solMap.addAll(li.next());
+		}
+		return solMap.iterator();
 	}
 
 	@Override
 	public boolean add( final SolutionMapping e ) {
-		// TODO Auto-generated method stub
-		return false;
+		final List<Node> valKeys = getVarKeys(e);
+
+		List<SolutionMapping> solMapList = map.get(valKeys);
+		if ( solMapList == null) {
+			solMapList = new ArrayList<>();
+			map.put(valKeys, solMapList);
+		}
+
+		solMapList.add(e);
+		return true;
 	}
 
 	@Override
 	public void clear() {
-		// TODO Auto-generated method stub
-		
+		final Iterator<List<SolutionMapping>> liValue = map.values().iterator();
+		while( liValue.hasNext() ){
+			liValue.next().clear();
+		}
+		map.clear();
 	}
 
 	@Override
 	public Iterable<SolutionMapping> getJoinPartners( final SolutionMapping sm )
 			throws UnsupportedOperationException
 	{
-		// TODO Auto-generated method stub
-		return null;
+		final List<Node> valKeys = getVarKeys(sm);
+		final Iterator<SolutionMapping> matchingSolMaps;
+		if (!valKeys.isEmpty() && valKeys.size() == joinVariables.size()){
+			final List<SolutionMapping> l = map.get(valKeys);
+			matchingSolMaps = (l != null) ? l.iterator() : null;
+		}
+		else if(valKeys.isEmpty()){
+			matchingSolMaps = iterator();
+		}
+		else {
+			throw new IllegalStateException();
+		}
+
+		final List<SolutionMapping> joinPartner = new ArrayList<>();
+		while (matchingSolMaps != null && matchingSolMaps.hasNext()) {
+			final SolutionMapping matchSolM = matchingSolMaps.next();
+			if (SolutionMappingUtils.compatible(sm, matchSolM)) {
+				joinPartner.add(matchSolM);
+			}
+		}
+		return joinPartner;
 	}
 
 	@Override
 	public Iterable<SolutionMapping> findSolutionMappings( final Var var, final Node value )
 			throws UnsupportedOperationException
 	{
-		// TODO Auto-generated method stub
-		return null;
+		if ( joinVariables.size() == 1 && joinVariables.contains(var) ){
+			final List<Node> valKeyL = new ArrayList<>();
+			valKeyL.add(value);
+			return map.get(valKeyL);
+		}
+		else{
+			throw new UnsupportedOperationException();
+		}
 	}
 
 	@Override
 	public Iterable<SolutionMapping> findSolutionMappings(
 			final Var var1, final Node value1,
 			final Var var2, final Node value2 )
-					throws UnsupportedOperationException
+			throws UnsupportedOperationException
 	{
-		// TODO Auto-generated method stub
-		return null;
+		if ( joinVariables.size() == 2 && joinVariables.contains(var1) && joinVariables.contains(var2) ){
+			final List<Node> valKeyL = new ArrayList<>();
+			for ( final Var v : joinVariables ) {
+				if( v.equals(var1) ){
+					valKeyL.add(value1);
+				}
+				else if( v.equals(var2) ){
+					valKeyL.add(value2);
+				}
+				else{
+					throw new IllegalArgumentException();
+				}
+			}
```

Comment:
Thanks for fixing this issue. Can you please add a unit test that uses this method and that would fail if this fix was not done. I mean, I see that you have a unit test that uses this method but in that unit test the order of the variables `var1` and `var2` is the same when creating the hash table and when probing into the hash table with this method here. Now, it would be good to have another version of this unit test where the order is swapped when probing.

Reply:
I added a unit test for testing this method. (commit ff86aa7)